lib/utils.js

```diff
@@ -36,7 +36,9 @@ exports.inherits = util.inherits;
  * @return {string}
  */
 exports.escape = function(html) {
-  return he.encode(String(html), {useNamedReferences: false});
+  return he
+    .encode(String(html), {useNamedReferences: false})
+    .replace(/&#x1B;/g, '');
```

Comment:
I think this is a too narrow solution. It can handle only the `#x1B;` case.
We can't replace special characters whenever we find invalid characters.

Reply:
From https://github.com/mochajs/mocha/issues/4526#issuecomment-737443533

> `Char ::= #x9 | #xA | #xD | [#x20-#xD7FF] | [#xE000-#xFFFD] | [#x10000-#x10FFFF]`

We could throw every hex char reference into a replacer function that checks against the spec's permitted range and strips invalid chars

```suggestion
    .replace(/&#x([0-9A-F]);/g, (match) => {
      const val = Number.parseInt(match); 
      return val === 0x9 || val === 0xA || val === 0xD || (val >= 0xE000 && val <= 0xFFFD) || (val >= 0x10000 && val <= 0x10FFFF) ? `&#x${match};` : '';
    });
```

or replace with a double escaped sequence:


```suggestion
    .replace(/&#x([0-9A-F]);/g, (match) => {
      const val = Number.parseInt(match); 
      return val === 0x9 || val === 0xA || val === 0xD || (val >= 0xE000 && val <= 0xFFFD) || (val >= 0x10000 && val <= 0x10FFFF) ? `&#x${match};` : `&amp;#x${match};`;
    });
```

Reply:
@outsideris do you have a preference?

Reply:
I prefer the first one. But I want to hear from @mochajs/core member's opinions because I don't have a specialty with xunit.

Reply:
@outsideris have you heard anything from the rest of the team?

Reply:
`utils.escape` is not only used by the `xunit` reporter, at least `HTML` reporter and maybe others would be affected. What are the implications?

> I think this is a too narrow solution [...]

I agree, we can't replace each special character step by step, this would be a never ending story.
If the `he` package can't help, there must be other packages available. I don't like the idea that Mocha has to handle special characters. What about eg. the `entities` package? Any others?

Reply:
https://github.com/mochajs/mocha/pull/4527#discussion_r551176184 has two proposals that handle the entire range of illegal characters, so it is not replacing "each special character step by step".

`entities` might provide a fix, but I wanted to avoid adding another dependency for something that is just a change of four lines of code.

https://github.com/michaelleeallen/mocha-junit-reporter as the most prominent external reporter uses `xml`, i.e. a proper XML parser, instead of just treating XML as a weird type of HTML.

Reply:
Coming back to this ~3 years later, I think it's fine to have a small hardcoded list in the source code. The list of characters doesn't change often and I'd rather avoid adding in another dependency.

I'm also in favor of stripping out unsafe characters (the first one, https://github.com/mochajs/mocha/pull/4527#discussion_r560348163). Double-escaping them makes the result somewhat unreadable IME.

Sorry to repeat the cycle of asking for more input, but cc @mochajs/maintenance-crew - thoughts?

If we don't get you an answer within, say, 3 weeks from now I say let's just go with your proposal of stripping out the unsafe characters. 😄

Reply:
> If we don't get you an answer within, say, 3 weeks from now I say let's just go with your proposal of stripping out the unsafe characters.

I'll take you up on this offer, @JoshuaKGoldberg and commit the fix that checks against the whole character range, but keeps stripping invalid chars.